### PR TITLE
175 forbid deleting running processes

### LINKED ISSUE #175: forbid deleting running processes on the backend
When deleteProcesses is called, trying to delete a running process should throw an error.

child oof ospin-web-dev/reactor#355

### PR DESCRIPTION
**Pull request by vdiezel**
closes #175 

First commit contains the logic changes to check for running processes and a small refactoring on the `delete` call in the process model.

Seconds commit contains refactoring of the test structure, including implementing the new seeder.
Most file changes due to moving files.

### CODE REVIEW

File: lib/validation-lib.js [hunk 1]
```diff
@@ -95,11 +95,11 @@ exports.isMyResource = (requestingUserId, targetUserId) => {
...
 
-exports.isNotPresent = (resources, id, label) => {
+exports.isNotPresent = (resources, id, label, errorMsg) => {
   if (resources.includes(id)) {
     return {
       invalid: true,
-      errorMsg: `The resource ${id} is already present in ${label}`,
+      errorMsg: errorMsg || `The resource ${id} is already present in ${label}`,
       responseFn: badRequest
     }
```
Comment (vdiezel) [inline lib/validation-lib.js]: provides the posisbility to pass in a custom message for speacial usage of the `isNotPresent` validator. Not too happy with this, but good for now.

File: services/process/src/process/process.js [hunk 2]
```diff
@@ -96,56 +96,80 @@ class Process {
...
-    for (let process of foundProcesses){
-      if (!targetDevicesIds.includes(process.deviceId)){
-        targetDevicesIds.push(process.deviceId);
-      }
-    }
 
-    const targetDevices = await this.devicesDB.find({ id: {$in: targetDevicesIds} }).toArray();
+    const targetDeviceIds = [
+      ...new Set(foundProcesses.map(foundProcess => foundProcess.deviceId))
+    ]
```
Comment (vdiezel) [inline services/process/src/process/process.js]: getting all unqiue deviceIds from the passed processes IDs

File: services/process/test/process/delete.test.js [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+          processesOnDevice1[0].id,
+          processesOnDevice2[0].id,
+        ]
+
+        // seed a bit of fake data that should remain
+        await dbSeeder.seedProcessFakeData(processesOnDevice2[4])
+
+        const remainingDataProcessId = processesOnDevice2[4].id
+
+        // grant user access to device2 and share a couple of processes with him
```
Comment (vdiezel) [inline services/process/test/process/delete.test.js]: `or her`

File: services/process/test/process/delete.test.js [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+          .find({ processId: { $in: deleteDataProcessIds }}).toArray()
+        expect(fetchedChunksDeleted.length).toBe(0)
+
+        // check if other process data remained
+
+        const fetchedChunksRemaining = await dbSeeder.dataPointsDB
+          .find({ processId: remainingDataProcessId }).toArray()
+        const upStreamCount = processesOnDevice2[4].deviceDescription
+          .filter(stream => stream.type === 'upstream').length
+        expect(fetchedChunksRemaining.length).toBe(upStreamCount * chunksPerStream)
```
Comment (vdiezel) [inline services/process/test/process/delete.test.js]: New: checking if data was deleted properly.

File: services/process/test/process/delete.test.js [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+        ]
+        const { res, body } = await fireDeleteProcessesEvent(allProcessIds, user.id)
+
+        expect(res.statusCode).toBe(400)
+        assertContainsAll(body.message, [ 'Bad Request', 'Running Process' ])
+
+        const processesCount = await dbSeeder.processesDB.count()
+        expect(processesCount).toBe(allProcessIds.length)
+      })
+    })
```
Comment (vdiezel) [inline services/process/test/process/delete.test.js]: New test covering the desired behaviour for this ticket: When trying to delete a running process, don't do anything and throw a 400

File: test/helper/Seeder.js [hunk 6]
```diff
@@ -93,14 +114,64 @@ class Seeder {
...
+  }
+
+  async seedProcessFakeData(process, chunksPerStream = 1) {
+    const dataChunks = process.deviceDescription
+      .filter(stream => stream.type === 'upstream')
+      .map(stream => Array.from({ length: chunksPerStream },
+        (_, i) => Seeder.generateFakeChunk(process, stream, i)))
+      .reduce((allChunks, currentChunks) => [ ...allChunks, ...currentChunks ], [])
+    await this.dataPointsDB.insertMany(dataChunks)
+  }
```
Comment (vdiezel) [inline test/helper/Seeder.js]: A couple more functions to make testing a bit easier.

File: services/process/src/process/process.js [hunk 7]
```diff
@@ -96,56 +96,80 @@ class Process {
...
+      .toArray()
 
-    for (let process of foundProcesses){
-      assert(() => hasObjectAccess(process, requestingUserId, RESOURCE.PROCESS, PROCESS_ACTIONS.DELETE))
-    }
+    foundProcesses.forEach(foundProcess => {
+      assert(() => hasObjectAccess(
+        foundProcess, requestingUserId, RESOURCE.PROCESS, PROCESS_ACTIONS.DELETE
+      ))
+    })
```
Comment (DanielSeehausen) [inline services/process/src/process/process.js]: 😎

File: services/process/src/process/process.js [hunk 8]
```diff
@@ -96,56 +96,80 @@ class Process {
...
-    const targetDevicesIds = [];
-    for (let process of foundProcesses){
-      if (!targetDevicesIds.includes(process.deviceId)){
-        targetDevicesIds.push(process.deviceId);
-      }
-    }
 
-    const targetDevices = await this.devicesDB.find({ id: {$in: targetDevicesIds} }).toArray();
+    const targetDeviceIds = [
+      ...new Set(foundProcesses.map(foundProcess => foundProcess.deviceId))
```
Comment (DanielSeehausen) [inline services/process/src/process/process.js]: love the set usage! very cool

File: services/process/src/process/process.js [hunk 9]
```diff
@@ -96,56 +96,80 @@ class Process {
...
+    })
 
-    await bulkUsers.execute();
+    await bulkUsers.execute()
 
-    const deletedProcesses = await this.processesDB.deleteMany({ id : { $in : processIdsToDelete }});
+    const deletedProcesses = await this.processesDB.deleteMany({ id: { $in: processIdsToDelete }})
 
     return deletedProcesses.deletedCount
   }
```
Comment (DanielSeehausen) [inline services/process/src/process/process.js]: as these methods grow, we should think about how we can break them up or at least separate them from one another. Nothing jumps to mind with how our current structure works, but wanted to float the idea for the future!

File: services/process/test/process/delete.test.js (hunk 3 above)
Comment (DanielSeehausen) [reply services/process/test/process/delete.test.js]: :laser:

File: services/process/test/process/delete.test.js (hunk 5 above)
Comment (DanielSeehausen) [reply services/process/test/process/delete.test.js]: https://stackoverflow.com/questions/13012842/which-http-status-code-to-return-when-the-delete-operation-is-not-allowed-for-pa

check this out!^

might be a more appropriate status code for this case

File: services/process/test/process/process.test.js [hunk 10]
```diff
@@ -4,30 +4,28 @@ const mongo = require('mongodb-lib')
...
+const revokeAccess = promisifyHandler(require('../../src/process/revokeAccess').main)
+const modifyAccess = promisifyHandler(require('../../src/process/modifyAccess').main)
+
+const { generateUser } = require('../../../user/test/seeds/user')
+const { generateDevice } = require('../../../device/test/seeds/device')
 const { generateProcess } = require('./seeds/process')
 const { addProcessesToDevice } = require('./utils/addProcessesToDevice')
 const { addProcessesToUser } = require('./utils/addProcessesToUser')
-const serializePublicProcess = require('../src/serializers/public/process')
+const serializePublicProcess = require('../../src/serializers/public/process')
```
Comment (DanielSeehausen) [inline services/process/test/process/process.test.js]: in the future we could add module loaders to avoid this like we have in REACTor !

**Comment by DanielSeehausen** [comment]
```diff
! forgot to mention - maybe there is a better status code to return (see comment). Not blocking though!
```

File: services/process/test/process/delete.test.js (hunk 5 above)
Comment (vdiezel) [reply services/process/test/process/delete.test.js]: great hint!

File: services/process/test/process/process.test.js (hunk 10 above)
Comment (vdiezel) [reply services/process/test/process/process.test.js]: would be great

**Pull request merged**